src/AggregateFunctions/AggregateFunctionOrderBy.h

```diff
@@ -0,0 +1,162 @@
+#include <Columns/ColumnArray.h>
+#include <Common/ArenaAllocator.h>
+#include <Common/assert_cast.h>
+#include <DataTypes/DataTypeArray.h>
+#include <Interpreters/AggregationCommon.h>
+
+namespace DB
+{
+
+    namespace ErrorCodes
+    {
+        extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    }
+
+struct AggregateFunctionOrderByData
+{
+    AggregateFunctionOrderByData() = default;
+    explicit AggregateFunctionOrderByData(size_t num_arguments): value(num_arguments), arena(nullptr) {}
+
+    using Array = std::vector<IColumn*>;
+
+    Array value;
+    std::once_flag initialized;
+
+    Arena* arena;
+};
+
+/** Adaptor for aggregate functions.
+  * Adding -OrderBy suffix to aggregate function
+**/
+
+class AggregateFunctionOrderBy final : public IAggregateFunctionDataHelper<AggregateFunctionOrderByData, AggregateFunctionOrderBy>
+{
+private:
+    AggregateFunctionPtr nested_func;
+    size_t num_arguments;
+    size_t prefix_size;
+    size_t num_sort_args;
+
+    AggregateDataPtr getNestedPlace(AggregateDataPtr place) const noexcept
+    {
+        return place + prefix_size;
+    }
+
+    ConstAggregateDataPtr getNestedPlace(ConstAggregateDataPtr place) const noexcept
+    {
+        return place + prefix_size;
+    }
+
+
+public:
+    AggregateFunctionOrderBy(AggregateFunctionPtr nested, const DataTypes & arguments, const Array & params)
+    : IAggregateFunctionDataHelper<AggregateFunctionOrderByData, AggregateFunctionOrderBy>(arguments, {params})
+    , nested_func(nested), num_arguments(arguments.size()), num_sort_args(params.back().safeGet<UInt64>())
+    {
+        prefix_size = sizeof(AggregateFunctionOrderByData);
+
+        if (arguments.empty())
+            throw Exception("Aggregate function " + getName() + " require at least one argument", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+    }
+
+    String getName() const override
+    {
+        return nested_func->getName() + "OrderBy";
+    }
+
+    DataTypePtr getReturnType() const override
+    {
+        return nested_func->getReturnType();
+    }
+
+    void create(AggregateDataPtr place) const override
+    {
+        new (place) AggregateFunctionOrderByData(num_arguments);
+        nested_func->create(getNestedPlace(place));
+    }
+
+    void destroy(AggregateDataPtr place) const noexcept override {
+        data(place).~AggregateFunctionOrderByData();
+        nested_func->destroy(getNestedPlace(place));
+    }
+
+    size_t sizeOfData() const override
+    {
+        return prefix_size + nested_func->sizeOfData();
+    }
+
+    size_t alignOfData() const override
+    {
+        return nested_func->alignOfData();
+    }
+
+    bool hasTrivialDestructor() const override
+    {
+        return nested_func->hasTrivialDestructor();
+    }
+
+    void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena * arena) const override
+    {
+        std::call_once(data(place).initialized, [this, place, columns]() {
```

Comment:
`add` is not supposed to be called concurrently on shared data, so the `call_once` shouldn't be necessary. Or is it?